pkg/ingest/conn_test.go

```diff
@@ -22,6 +22,112 @@ import (
 	"github.com/oklog/oklog/pkg/fs"
 )
 
+func TestDynamicRecordReader(t *testing.T) {
+	cases := []struct {
+		err   error
+		input string
+		exp   []string
+	}{
+		{
+			input: "topic_A foo1\ntopic_B foo2\ntopic_A foo2\n",
+			exp:   []string{"topic_A foo1\n", "topic_B foo2\n", "topic_A foo2\n"},
+		},
+		// No final newline. Final record is considered torn and skipped.
+		{
+			input: "topic_A foo1\ntopic_B foo2\ntopic_A foo2",
+			exp:   []string{"topic_A foo1\n", "topic_B foo2\n"},
+		},
+		// No delimited topic identifier.
+		{
+			input: "topic_A foo1\nabcdef\ntopic_B foo2\n",
+			exp:   []string{"topic_A foo1\n"},
+			err:   ErrIllegalTopicFormat,
+		},
+		// Empty record in the middle.
+		{
+			input: "topic_A foo1\n\ntopic_B foo2\n",
+			exp:   []string{"topic_A foo1\n"},
+			err:   ErrIllegalTopicFormat,
+		},
+		// Bad topic characters.
+		{
+			input: "to~pic foo1\n",
+			err:   ErrIllegalTopicFormat,
+		},
+	}
```

Comment:
Tip: if you give the test case struct a `name string`, and move the descriptions from comments to that name, you can do

```go
for _, c := range []struct{
    name  string
    input string
    want  []string
    err   error
}{
    {
        name:  "basic",
        input: "...",
        want:  []string{...},
    },
}{
    t.Run(c.name, func(t *testing.T) {
        // subtest here
    })
}
```

Reply:
Ah, neat! 👍